tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -3116,7 +3163,23 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
       !IsUnrollingColumnReductionBeneficial(unnested_hlo, input_shape,
                                             reduction_dimensions.dimensions[2]);
 
-  if (!dilated_x && !reduction_dimensions.is_row_reduction) {
+  KernelMappingScheme::IndexingOrder indexing_order;
+  if (reduction_dimensions.is_row_reduction &&
+      // Only try to vectorize+coales memory access for row of even size.
+      // For odd row size, every other row isn't aligned, so can't be
+      // vectorized.
+      reduction_dimensions.dimensions[2] % 2 == 0) {
+    indexing_order = kLinearStridedIndexingX;
+  } else if (IsUnrollingColumnReductionBeneficial(
+                 unnested_hlo, input_shape,
+                 reduction_dimensions.dimensions[2])) {
+    indexing_order = kLinearIndexingX;
+  } else {
+    indexing_order = kStridedIndexingX;
+  }
+
+  if (indexing_order == kLinearIndexingX &&
+      !reduction_dimensions.is_row_reduction) {
     // Vectorized loads: a single thread reduces two adjacent columns.
```

Comment:
But not all row reductions are vectorized?

Reply:
I'm not sure I understand your question.
This comment was already in the code in the past.
This path is for column reduction, where each thread reduce 2 consecutive output value at the same time.

row reduction aren't always vectorized. When the tile fit, they are always vectorized. For even row size, it is partially vectorized. The missing vectorized is due to a problem of LLVM optimization pass order.
For odd row size, I do not trigger the new indexing order as it doesn't get vectorized by LLVM and this slow down my benchmark.